packages/run-protocol/src/feeDistributor.js

```diff
@@ -0,0 +1,302 @@
+      return E(collectFeesSeat).getPayout('RUN');
+    },
+  });
+}
+
+/**
+ * A distributor of fees from Inter Protocol sources to deposit facets. Each
+ * time the epochTimer signals the end of an Epoch, it will ask the contracts
+ * for fees that have been collected to date and send that payment to the
+ * depositFacet.
+ *
+ * @param {() => Promise<any>} schedulePayments - distribute to the destinations
+ * @param {ERef<TimerService>} timerService - timer that is used to schedule collections
+ * @param {RelativeTime} [collectionInterval] - how often to collect fees
+ */
+export function startDistributing(
+  schedulePayments,
+  timerService,
+  collectionInterval = 1n,
+) {
+  const timeObserver = {
+    updateState: _ =>
+      schedulePayments().catch(e => {
+        console.error('failed with', e);
+        throw e;
+      }),
+    fail: reason => {
+      throw Error(`fee distributor timer failed: ${reason}`);
+    },
+    finish: done => {
+      throw Error(`fee distributor timer died: ${done}`);
+    },
+  };
+
+  const collectionNotifier = E(timerService).makeNotifier(
+    0n,
+    collectionInterval,
+  );
+  observeNotifier(collectionNotifier, timeObserver).catch(e => {
+    console.error('fee distributor failed with', e);
+  });
+}
+
+/**
+ * @type {BuildFeeDistributor}
+ */
+export const buildDistributor = async (
+  collectors,
+  depositFacet,
+  timerService,
+  { epochInterval = 1n },
+) => {
+  startDistributing(
+    () =>
+      Promise.all(
+        collectors.map(collector =>
+          E(collector)
+            .collectFees()
+            .then(pmt => E(depositFacet).receive(pmt)),
+        ),
+      ),
+    timerService,
+    epochInterval,
+  );
+};
+
+/**
+ * @typedef {{ pushPayment: (payment: Payment, issuer: Issuer) => Promise<Amount>}} FeeDestination
+ *
+ * @param {Record<string, ERef<FeeDestination>>} [destinations]
+ * @param {Record<string, any>} [keywordShares]
+ */
+export const makeShareConfig = (destinations = {}, keywordShares = {}) => {
+  let totalShares = 0n;
+  const shares = Object.entries(destinations)
+    .filter(([_, dst]) => dst)
+    .map(([kw, destination]) => {
+      const share = keywordShares[kw];
+      assert.typeof(share, 'bigint', `${kw} must be a bigint; got ${share}`);
+      totalShares += share;
+      return {
+        share,
+        destination,
+      };
+    });
+  return { shares, totalShares };
+};
+
+/** @typedef {ReturnType<typeof makeShareConfig>} ShareConfig */
+
+/**
+ * shareFees tries to be tolerant of single points of failure.
+ *
+ * @param {Payment} payment
+ * @param {Issuer} issuer
+ * @param {ShareConfig} shareConfig
+ */
+export const shareFees = async (payment, issuer, { shares, totalShares }) => {
+  // Divide the payment into shares.
+  /** @type {Amount<'nat'>} */
+  const totalAmount = await E(issuer).getAmountOf(payment);
+  if (AmountMath.isEmpty(totalAmount)) {
+    return;
+  }
```

Comment:
consider also returning if there are no shares. It's not immediately obvious that this does nothing in that case.